exonum-java-binding/qa-service/src/test/java/com/exonum/binding/qaservice/QaServiceImplIntegrationTest.java

```diff
@@ -114,356 +97,401 @@ void tearDown() {
+        int expectedNumMessages = 2;
+        assertThat(logMessages).hasSize(expectedNumMessages);
+
+        assertThat(logMessages.get(0))
+            .contains("ERROR")
+            .contains(stateHashes.get(0).toString());
+        return null;
+      });
     }
   }
 
   @Test
-  void submitCreateCounter() throws Exception {
-    setServiceNode(node);
-
-    when(node.submitTransaction(any(RawTransaction.class))).thenReturn(DEFAULT_TX_MESSAGE_HASH);
-
-    String counterName = "bids";
-    HashCode txHash = service.submitCreateCounter(counterName);
-
-    CreateCounterTx expectedTx = new CreateCounterTx(counterName);
-    RawTransaction expectedRawTx = CreateCounterTx.converter().toRawTransaction(expectedTx);
-
-    assertThat(txHash).isEqualTo(DEFAULT_TX_MESSAGE_HASH);
-    verify(node).submitTransaction(eq(expectedRawTx));
+  void initialize() {
+    try (TestKit testKit = TestKit.forService(QaServiceModule.class)) {
+      testKit.withSnapshot((view) -> {
+        // TODO: https://jira.bf.local/browse/ECR-2683 is needed for service configuration to be
+        //  checked
+
+        // Check that both the default and afterCommit counters were created.
+        QaSchema schema = new QaSchema(view);
+        MapIndex<HashCode, Long> counters = schema.counters();
+        MapIndex<HashCode, String> counterNames = schema.counterNames();
+
+        HashCode defaultCounterId = sha256().hashString(DEFAULT_COUNTER_NAME, UTF_8);
+        HashCode afterCommitCounterId = sha256().hashString(AFTER_COMMIT_COUNTER_NAME, UTF_8);
+
+        assertThat(counters.get(defaultCounterId)).isEqualTo(0L);
+        assertThat(counterNames.get(defaultCounterId)).isEqualTo(DEFAULT_COUNTER_NAME);
+        assertThat(counters.get(afterCommitCounterId)).isEqualTo(0L);
+        assertThat(counterNames.get(afterCommitCounterId)).isEqualTo(AFTER_COMMIT_COUNTER_NAME);
+        return null;
+      });
+    }
   }
 
   @Test
-  void submitIncrementCounter() throws Exception {
-    setServiceNode(node);
-    when(node.submitTransaction(any(RawTransaction.class))).thenReturn(DEFAULT_TX_MESSAGE_HASH);
-
-    long seed = 1L;
-    HashCode counterId = sha256()
-        .hashString("Cats counter", UTF_8);
-    HashCode txHash = service.submitIncrementCounter(seed, counterId);
-
-    IncrementCounterTx expectedTx = new IncrementCounterTx(seed, counterId);
-    RawTransaction expectedRawTx = IncrementCounterTx.converter().toRawTransaction(expectedTx);
-
-    assertThat(txHash).isEqualTo(DEFAULT_TX_MESSAGE_HASH);
-    verify(node).submitTransaction(eq(expectedRawTx));
+  void submitCreateCounter() {
+    try (TestKit testKit = TestKit.forService(QaServiceModule.class)) {
+      QaServiceImpl service = testKit.getService(QaService.ID, QaServiceImpl.class);
+      String counterName = "bids";
+      service.submitCreateCounter(counterName);
+      List<TransactionMessage> inPoolTransactions =
+          testKit.findTransactionsInPool(tx -> tx.getServiceId() == QaService.ID
```

Comment:
(Here and below) These verifications became rather loose (they don't check the payload).

What is the easiest way to verify that a service submitted a particular transaction _properly_? Construct an expected message (but it requires the service key pair)? An expected 'RawTransaction'? Convert the message into Raw and decode? Can testkit provide anything to simplify this?

Reply:
I updated the tests so we construct an expected transaction message. We could provide transaction message construction in TestKit - given raw transaction it will sign it with emulated node key pair.

Reply:
Let's ponder on that in a separate issue: https://jira.bf.local/browse/ECR-3189